share/osutil/file_linux.go

```diff
@@ -42,14 +43,72 @@ type FileInfoExt struct {
 	UserAdded   bool
 }
 
+func extractFilePath(symlinkPath, errorMessage, delimiter string) string {
+	// Split the error message into two parts 
+	// TODO => make sure we get the file out only if the read link
+	parts := strings.Split(errorMessage, delimiter)
+	if len(parts) < 2 {
+		return ""
+	}
+
+	// Extract the command and file path part
+	commandFilePath := parts[0]
+
+	// Further split by space and take the last element
+	pathParts := strings.Split(commandFilePath, " ")
+	if len(pathParts) == 0 {
+		return ""
+	}
+
+	resultPath := pathParts[len(pathParts)-1]
+	relativePath, err := filepath.Rel(filepath.Dir(symlinkPath), resultPath)
+
+    if err != nil {
+        fmt.Println("Error:", err)
+        return ""
+    }
+
+	return relativePath
+}
+
+func fileExists(path string) bool {
+    _, err := os.Stat(path)
+    return !errors.Is(err, os.ErrNotExist)
+}
+
+func resolvePath(symlinkPath, errMsg string) (string, error) {
+	symlink := extractFilePath(symlinkPath, errMsg, ":")
+    root := strings.Join(strings.Split(symlinkPath, "/")[:4], "/") 
+
+	parts := strings.Split(symlink, "/")
+	for i := range parts {
+		// Join the components from the current index to the end
+		partialSymlink := strings.Join(parts[i:], "/")
+		resolvedPath := filepath.Join(filepath.Dir(symlinkPath), partialSymlink)
+		
+		if strings.HasPrefix(resolvedPath, root) && fileExists(resolvedPath) {
+			return resolvedPath, nil 
+		}
+	}
+
+
+    return "", errors.New("invalid file path")
+}
+
 //try to get sym link
 func GetContainerRealFilePath(pid int, path string) (string, error) {
+	log.WithFields(log.Fields{"pid": pid, "path": path}).Info("in GetContainerRealFilePath")
 	retry := 0
+	symlinkPath := path
 	for retry < linkMaxLayers {
 		linkPath, err := os.Readlink(path)
 		if err != nil {
-			log.WithFields(log.Fields{"error": err}).Debug("Read file link fail")
-			return "", err
+			if resolvedPath, resolveErr := resolvePath(symlinkPath, err.Error()); resolveErr != nil{
+				log.WithFields(log.Fields{"error": err}).Debug("Read file link fail")
+				return "", err
+			} else {
+				return resolvedPath, nil
+			}
```

Comment:
You shouldn't only rely on whether the file exists to determine if it's a correct path. You need to examine its path to make sure it doesn't escape from `/proc/*/root`. For example, the file pointed by this symlink would exist, but it's not correct.  (It will find host's `/etc/passwd` instead of container's. )

```
ln -s ../../../../../etc/passwd /bin/rpm
```